### PR TITLE
Support under Windows/MSYS

### PR DESCRIPTION
**Pull request by matrixji**
issue: #131
ref: https://github.com/milvus-io/knowhere/issues/131

Signed-off-by: Ji Bin <matrixji@live.com>

### CODE REVIEW

**Comment by cydrain** [comment]
/hold

Hi @matrixji , thanks a lot for your contribute for knowhere.
We are upgrading knowhere from faiss-1.7.0 to latest faiss-1.7.2 to support Mac OS.
I'll hold your PR till the upgrade has finished.

**Comment by matrixji** [comment]
> It looks like this is your first PR to milvus-io/knowhere

Good news for upgrading 1.7.2, seems fewer workarounds are needed for MSYS, I'll rebase after that.

**Comment by cydrain** [comment]
Hi @matrixji , we are refactoring knowhere cmake, we'll let you know when it's done.

**Comment by matrixji** [comment]
> Hi @matrixji , we are refactoring knowhere cmake, we'll let you know when it's done.

I see PR for using knowhere/master is raised in `milvus-io/milvus`, as [Support compile under windows](https://github.com/milvus-io/milvus/pull/15786) is merged. I guess this PR may be needed.

**Comment by matrixji** [comment]
rebased due to conflict with the latest main branch.

**Comment by cydrain** [comment]
Hi @matrixji ,
It's delayed to refactor knowhere cmake, please resolve these conflicts, I can merge your code in.
And since it may take a few days to merge your code in, can you help to upgrade v1.1.1-windows-adaptation.tar.gz in your repo to knowhere-v1.1.2 to let Milvus "Code Checker MinGW" pass ?
Thanks a lot.

**Comment by matrixji** [comment]
Rebased.

**Comment by matrixji** [comment]
> 

Hi @cydrain,
Rebased, now you could review those changes.

File: CMakeLists.txt [hunk 1]
```diff
@@ -133,6 +135,10 @@ if ( LINUX )
     append_flags( CMAKE_CXX_FLAGS FLAGS "-Werror" "-rdynamic" )
 endif ()
 
+if ( MSYS )
```
Comment (cydrain) [inline CMakeLists.txt:138]: you can use "elseif"

File: knowhere/index/vector_index/ConfAdapterMgr.cpp [hunk 2]
```diff
@@ -15,6 +15,12 @@
 
 namespace knowhere {
 
+AdapterMgr&
```
Comment (cydrain) [inline knowhere/index/vector_index/ConfAdapterMgr.cpp:18]: these changes are not related with Windows-support, please keep it untouched.

File: knowhere/index/vector_index/ConfAdapterMgr.h [hunk 3]
```diff
@@ -31,10 +31,7 @@ class AdapterMgr {
     };
 
     static AdapterMgr&
-    GetInstance() {
```
Comment (cydrain) [inline knowhere/index/vector_index/ConfAdapterMgr.h:34]: these changes are not related with Windows-support, please keep it untouched.

File: knowhere/index/vector_index/helpers/SPTAGParameterMgr.cpp [hunk 4]
```diff
@@ -23,6 +23,12 @@ SPTAGParameterMgr::GetBKTParameters() {
     return bkt_config_;
 }
 
+SPTAGParameterMgr&
```
Comment (cydrain) [inline knowhere/index/vector_index/helpers/SPTAGParameterMgr.cpp:26]: these changes are not related with Windows-support, please keep it untouched.

File: knowhere/index/vector_index/helpers/SPTAGParameterMgr.h [hunk 5]
```diff
@@ -32,10 +32,7 @@ class SPTAGParameterMgr {
 
  public:
     static SPTAGParameterMgr&
-    GetInstance() {
```
Comment (cydrain) [inline knowhere/index/vector_index/helpers/SPTAGParameterMgr.h:35]: these changes are not related with Windows-support, please keep it untouched.

File: knowhere/utils/BitsetView.h [hunk 6]
```diff
@@ -214,7 +214,7 @@ class ConcurrentBitset {
             return x;
         };
         for (size_t i = 0; i < len; ++i) {
-            ret += __builtin_popcountl(*p_data);
```
Comment (cydrain) [inline knowhere/utils/BitsetView.h:217]: we are rewriting BitsetView, you can skip this new

File: knowhere/utils/BitsetView.h [hunk 7]
```diff
@@ -308,7 +308,7 @@ class BitsetView {
             return x;
         };
         for (int64_t i = 0; i < len; ++i) {
-            ret += __builtin_popcountl(*p_data);
```
Comment (cydrain) [inline knowhere/utils/BitsetView.h:311]: we are rewriting BitsetView, you can skip this now

File: scripts/install_deps.sh [hunk 8]
```diff
@@ -80,3 +80,14 @@ fi
...
     brew install boost libomp llvm ninja tbb openblas
 fi
+
+if [[ "${MACHINE}" == "MinGw"  ]]; then
+    pacman -Su --noconfirm --needed \
+    git make tar dos2unix zip unzip patch \
+    mingw-w64-x86_64-toolchain \
+    mingw-w64-x86_64-make \
+    mingw-w64-x86_64-cmake \
+    mingw-w64-x86_64-boost \
```
Comment (cydrain) [inline scripts/install_deps.sh:90]: knowhere has no dependency on boost and tbb, I have remove boost and tbb for ubuntu, centos and mac

File: thirdparty/CMakeLists.txt [hunk 9]
```diff
@@ -105,6 +105,14 @@ macro(build_faiss)
                  --without-cuda )
     endif ()
 
+    if (MSYS)
+        # patch for folder faiss/faiss, symlink not work for msys
+        execute_process(
+            COMMAND bash -c "test -f faiss/faiss/Index.h || (rm -fr faiss/faiss && cp -fr faiss faiss.tmp && mv faiss.tmp faiss/faiss)"
```
Comment (cydrain) [inline thirdparty/CMakeLists.txt:111]: for faiss-1.7.2, faiss/faiss is not a symbol link any more

File: thirdparty/NGT/lib/NGT/MmapManagerImpl.hpp [hunk 10]
```diff
@@ -20,7 +20,11 @@
 #include "MmapManagerDefs.h"
 #include "MmapManagerException.h"
 
+#ifndef __MINGW64__
 #include <sys/mman.h>
+#else
+#include "../../annoy/src/mman.h"
```
Comment (cydrain) [inline thirdparty/NGT/lib/NGT/MmapManagerImpl.hpp:26]: not reasonable to let NGT include annoy header file

File: thirdparty/faiss/faiss/impl/index_read.cpp [hunk 11]
```diff
@@ -14,7 +14,11 @@
 #include <cstdio>
 #include <cstdlib>
 
+#ifdef __MINGW64__
+#include "../annoy/src/mman.h"
```
Comment (cydrain) [inline thirdparty/faiss/faiss/impl/index_read.cpp:18]: don't let faiss include annoy header file

File: thirdparty/faiss/faiss/invlists/OnDiskInvertedLists.cpp [hunk 12]
```diff
@@ -13,7 +13,11 @@
 
 #include <unordered_set>
 
+#ifdef __MINGW64__
+#include "../annoy/src/mman.h"
```
Comment (cydrain) [inline thirdparty/faiss/faiss/invlists/OnDiskInvertedLists.cpp:17]: don't let faiss include annoy header file

File: thirdparty/faiss/faiss/invlists/OnDiskInvertedLists.cpp [hunk 13]
```diff
@@ -523,7 +527,7 @@ void OnDiskInvertedLists::free_slot(size_t offset, size_t capacity) {
         it++;
     }
 
-    size_t inf = 1UL << 60;
```
Comment (cydrain) [inline thirdparty/faiss/faiss/invlists/OnDiskInvertedLists.cpp:526]: need not change this, both 1UL and 1ULL are 8 bytes.

File: unittest/test_binaryidmap.cpp [hunk 14]
```diff
@@ -118,7 +118,7 @@ TEST_P(BinaryIDMAPTest, binaryidmap_serialize) {
         auto binaryset = index_->Serialize(conf);
         auto bin = binaryset.GetByName("BinaryIVF");
 
-        std::string filename = "/tmp/bianryidmap_test_serialize.bin";
```
Comment (cydrain) [inline unittest/test_binaryidmap.cpp:121]: not necessary

File: thirdparty/faiss/faiss/invlists/OnDiskInvertedLists.cpp (hunk 13 above)
Comment (matrixji) [reply thirdparty/faiss/faiss/invlists/OnDiskInvertedLists.cpp:526]: In windows, UL means `unsigned long` which is 4bytes(32bits) even in win64. 
Ref: https://stackoverflow.com/questions/64662711/c-64bit-compilation-on-vs-code-with-windows-10

File: thirdparty/faiss/faiss/invlists/OnDiskInvertedLists.cpp (hunk 12 above)
Comment (matrixji) [reply thirdparty/faiss/faiss/invlists/OnDiskInvertedLists.cpp:17]: Currently, we need this `mman` for support mmap under windows, how about we copy it to some dedicated common folder?

File: scripts/install_deps.sh (hunk 8 above)
Comment (matrixji) [reply scripts/install_deps.sh:90]: I'll remove it in the next rebased version.

File: knowhere/index/vector_index/ConfAdapterMgr.cpp (hunk 2 above)
Comment (matrixji) [reply knowhere/index/vector_index/ConfAdapterMgr.cpp:18]: Unfortunately, the `implementing singleton in the header has different` behaviors between Linux/GCC and MinGW/GCC.
In MinGW, it will cause create multiple instances, but it does right under Linux (only one instance). I could find some suitable references to explain why so fat, for my understanding, the root cause may be the diff Relocation details between PE and ELF format.

File: knowhere/utils/BitsetView.h (hunk 6 above)
Comment (matrixji) [reply knowhere/utils/BitsetView.h:217]: I see the PR https://github.com/milvus-io/knowhere/pull/125 ,  note: seems __builtin_popcountl is still used in the PR and it's a 32bit function in Windows/MinGW.

File: thirdparty/CMakeLists.txt (hunk 9 above)
Comment (matrixji) [reply thirdparty/CMakeLists.txt:111]: I'll remove it in the next rebased version.

File: knowhere/utils/BitsetView.h (hunk 6 above)
Comment (cydrain) [reply knowhere/utils/BitsetView.h:217]: I have tested under ubuntu, both 1UL and 1ULL occupy 8 bytes.
Since __builtin_popcountl accept 32bit in MinGW, I'll replace it with __builtin_popcountll

File: thirdparty/NGT/lib/NGT/MmapManagerImpl.hpp (hunk 10 above)
Comment (cydrain) [reply thirdparty/NGT/lib/NGT/MmapManagerImpl.hpp:26]: If you have no better way to resolve this issue, you can skip NSG and SPTAG.
These 2 indices are banned now since few customers are using them.

**Comment by matrixji** [comment]
Hi, @cydrain, rebased. And should fixed all issues from your comments.

**Comment by matrixji** [comment]
@mergify refresh

**Comment by matrixji** [comment]
Hi @cydrain,  I just update the commit body for `dco-passed`, needs `lgtm` and `approve` again.

**Comment by matrixji** [comment]
Oops, buy current mergify settings, seems` do-not-merge/missing-related-issue` label will not be removed when the status updated.
